Source/DafnyLanguageServer/Language/ImplicitFailingAssertionCodeActionProvider.cs

```diff
@@ -0,0 +1,117 @@
+  }
+
+  protected static List<Node>? FindInnermostNodeIntersecting(Node node, Range range) {
+    if (node.RangeToken.StartToken.line > 0 && !node.RangeToken.ToLspRange().Intersects(range)) {
+      return null;
+    }
+
+    foreach (var child in node.PreResolveChildren) {
+      var result = FindInnermostNodeIntersecting(child, range);
+      if (result != null) {
+        result.Add(node);
+        return result;
+      }
+    }
+
+    return new List<Node>() { node };
+  }
+
+  class ExplicitAssertionDafnyCodeAction : DafnyCodeAction {
+    private readonly DafnyOptions options;
+    private readonly Expression failingImplicitAssertion;
+    private readonly Node program;
+    private readonly Range selection;
+
+    public ExplicitAssertionDafnyCodeAction(
+      DafnyOptions options,
+      Dafny.Program program,
+      Expression failingImplicitAssertion,
+      Range selection
+      ) : base("Explicit failing assertion") {
+      this.options = options;
+      this.failingImplicitAssertion = failingImplicitAssertion;
+      this.program = program;
+      this.selection = selection;
+    }
+
+    public override IEnumerable<DafnyCodeActionEdit> GetEdits() {
+      var nodesTillFailure = FindInnermostNodeIntersecting(program, selection);
+
+      var suggestedEdits = new List<DafnyCodeActionEdit>();
+      if (nodesTillFailure != null) {
+        Node? insertionNode = null;
+        for (var i = 0; i < nodesTillFailure.Count; i++) {
+          var node = nodesTillFailure[i];
+          var nextNode = i < nodesTillFailure.Count - 1 ? nodesTillFailure[i + 1] : null;
+          if ((node is Statement or LetExpr && (node is not UpdateStmt || nextNode is not VarDeclStmt))
+              || nextNode is TopLevelDecl or MemberDecl or ITEExpr or MatchExpr or NestedMatchExpr or NestedMatchCase
+             ) {
+            insertionNode = node;
+            break;
+          }
+        }
+
+        if (insertionNode == null) {
+          insertionNode = nodesTillFailure[0];
+        }
+
+        var start = insertionNode.StartToken;
+        var assertStr = $"assert {Printer.ExprToString(options, failingImplicitAssertion)};\n" +
+                        IndentationFormatter.Whitespace(Math.Max(start.col - 1, 0));
+        suggestedEdits.Add(
+          new DafnyCodeActionEdit(
+            new RangeToken(start, null), assertStr));
+      }
+
+      return suggestedEdits.ToArray();
+    }
+  }
+
+  protected override IEnumerable<DafnyCodeAction>? GetDafnyCodeActions(IDafnyCodeActionInput input,
+    DafnyDiagnostic diagnostic, Range selection) {
+    var uri = input.Uri;
+    if (input.Program == null || diagnostic.Source != MessageSource.Verifier) {
+      return null;
+    }
+    var failingExpressions = new List<Expression>() { };
+    input.VerificationTree.ForEach(tree => {
+      if (tree is AssertionVerificationTree assertTree &&
+          assertTree.Finished &&
+            assertTree.Range.Intersects(selection) &&
```

Comment:
This might match failed implicit assertions located in the specification text, right? Will it insert invalid syntax in that case?

This might work fine if you are able to insert `assert ...; <expr>` there, but we should at least have some test cases for that.

Reply:
It will work fine in that case because it will do exactly what you suggest, that is, insert `assert ...;` before <expr>, even in specification expressions